src/workerd/api/sockets.h

```diff
@@ -10,83 +10,24 @@
-  // are deferred until the `inner` promise completes.
-public:
-  explicit PipelinedAsyncIoStream(kj::Promise<kj::Own<kj::AsyncIoStream>> inner);
-  void shutdownWrite() override;
-  void abortRead() override;
-  void getsockopt(int level, int option, void* value, uint* length) override;
-  void setsockopt(int level, int option, const void* value, uint length) override;
-
-  void getsockname(struct sockaddr* addr, uint* length) override;
-  void getpeername(struct sockaddr* addr, uint* length) override;
-
-  // AsyncInputStream methods:
-  kj::Promise<size_t> read(void* buffer, size_t minBytes, size_t maxBytes) override;
-  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
-
-  // AsyncOutputStream methods:
-  kj::Promise<void> write(const void* buffer, size_t size) override;
-  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override;
-
-  kj::Promise<void> whenWriteDisconnected() override;
-
-private:
-  template <typename T>
-  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
-    // Either calls `then` on the `inner` promise with the `f` callback, or calls `f` on an already
-    // stored stream (if the `inner` promise completed).
-    KJ_IF_MAYBE(e, error) {
-      kj::throwRecoverableException(kj::mv(*e));
-    }
-
-    KJ_IF_MAYBE(s, completedInner) {
-      return f(s);
-    } else {
-      return inner.then([this, f](kj::Own<kj::AsyncIoStream> stream) -> kj::Promise<T> {
-        completedInner = kj::mv(stream);
-        return f(&KJ_ASSERT_NONNULL(completedInner));
-      });
-    }
-  }
-
-  kj::Maybe<kj::Own<kj::AsyncIoStream>> completedInner;
-  // Stored io stream once the `inner` promise completes.
-  kj::Maybe<kj::Exception> error;
-  // Stored error if any of the operations throw.
-  kj::Promise<kj::Own<kj::AsyncIoStream>> inner;
-  // A promise holding a stream that will be available at some point in the future. Typically once
-  // a connection is established.
-};
-
 struct SocketOptions {
   jsg::Unimplemented tls; // TODO(later): TCP socket options need to be implemented.
   JSG_STRUCT(tls);
 };
 
-struct InitData {
-  jsg::Ref<ReadableStream> readable;
-  jsg::Ref<WritableStream> writable;
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
-};
-
 class Socket: public jsg::Object {
 public:
-  Socket(InitData data)
-      : readable(kj::mv(data.readable)), writable(kj::mv(data.writable)),
-        closeFulfiller(kj::mv(data.closeFulfiller)) {};
-  Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise);
+  Socket(jsg::Lock& js, jsg::Ref<ReadableStream> readable, jsg::Ref<WritableStream> writable,
+      kj::Own<jsg::PromiseResolverPair<void>> close)
+      : readable(kj::mv(readable)), writable(kj::mv(writable)),
+        closeFulfiller(kj::mv(close)),
+        closedPromise(kj::mv(closeFulfiller->promise)) {
```

Comment:
One other bit here. Currently, the way this is set up, if the `closedPromise` rejects and there is no one handling rejections on the `socket.closed` property, an `unhandledrejection` event will occur. I'm not convinced that's ideal by default. We should probably mark the `closedPromise` as handled explicitly so that the unhandled rejection event only occurs if someone attaches a `then` handler to the promise without a corresponding `catch` handler.

Reply:
I suppose now that we have `close` returning a promise again we can instead reject that returned promise instead of the `closed` promise (and I see below that this is what Kenton is suggesting we should do). So I'll do that and that will resolve this issue too.

Reply:
Btw can you elaborate on what "marking the `closedPromise` as handled explicitly" means? how do I do that?

Reply:
See the `markAsHandled()` method on `jsg::Promise` (https://github.com/cloudflare/workerd/blob/002cf0d582e9effd6bad46461c4ca4f242579563/src/workerd/jsg/promise.h#L232)

Reply:
Okay, I added it where I construct the `PromiseResolverPair` (couldn't figure out how to access it through the `jsg::MemoizedIdentity`)